content/kb/troubleshooting-failure-to-delete-orphaned-pod-volume-directory.md

```diff
@@ -0,0 +1,35 @@
+---
+title: "Troubleshooting: Failure to delete orphaned pod volume directory"
+author: Ray Chang
+draft: false
+date: 2023-03-23
+categories:
+  - "csi"
+---
+
+## Applicable versions
+
+All Longhorn versions.
+
+## Symptoms
+
+When a worker node crashes with some pods running on it, many error messages as shown below keep appearing in the kubelet.
+
+```
+orphaned pod <pod-uid> found, but error not a directory occurred when trying to remove the volumes dir
+```
+
+## Reason
+
+This is caused by the kubelet trying to tear down the volume by sending a CSI request to the longhorn-csi-plugin, but a node failure resulted in a disconnect. Then, when the kubelet tries to delete `vol_data.json` to unmount the volume, it also failed due to the failure of the CSI request.
```

Comment:
IMO, this statement does not clarify the issue...
Isn't the root cause that the **invalid mount point** caused by the node crash would fail the **volume reconstruction** (upstream bug) during the node/kubelet recovery?

Reply:
@shuo-wu Could you link me to the document describing this root cause? 

From the comment of @derekbit, it is mentioning that the root cause is the disconnection between kubelet and longhorn-csi-plugin https://github.com/longhorn/longhorn/issues/3207#issuecomment-974845277. So kubelet cannot tell longhorn-csi-plugin to clean up that stale `volume_data.json` file

Reply:
Oh sorry, I didn't notice this explanation. My opinion on this issue is from the upstream fix: It removes the existence check of mount point for filesystem volume or symbolic link for block volume.
But here @derekbit confirmed that it's caused by the CSI plugin connection breaking... Not sure which one is the root

Reply:
I feel both are the same thing, but the node hard reboot/crash is just one case of `the disconnection between kubelet and longhorn-csi-plugin`.

@weizhe0422 can base on the reviews here fine-tune the PR.

Reply:
I reference to https://github.com/kubernetes/kubernetes/issues/111933 and https://github.com/kubernetes/kubernetes/pull/116134 and think the situation is lead caused by the connection broken between kubelet and plug-in, and the kubelet will try to reconstruct when resuming but failed since the volume mount point on the reboot node is out of date. 
https://github.com/kubernetes/kubernetes/blob/b2f26460b9354b38b174a21cdbf15c391d63dade/pkg/kubelet/volumemanager/reconciler/reconstruct_common.go#L296-L303
```golang
isExist, checkErr := rc.operationExecutor.CheckVolumeExistenceOperation(volumeSpec, checkPath, volumeSpec.Name(), rc.mounter, uniqueVolumeName, volume.podName, pod.UID, attachablePlugin)
if checkErr != nil {
            return nil, checkErr
}
// If mount or symlink doesn't exist, volume reconstruction should be failed
if !isExist {
            return nil, fmt.Errorf("volume: %q is not mounted", uniqueVolumeName)
 }
```
Therefore, the mount point information cannot be properly updated so that it is cleaned up in the next reconcile cycle.
https://github.com/kubernetes/kubernetes/blob/16534deedf1e3f7301b20041fafe15ff7f178904/pkg/kubelet/volumemanager/reconciler/reconstruct_new.go#L155-L174
```golang
func (rc *reconciler) cleanOrphanVolumes() {
            if len(rc.volumesFailedReconstruction) == 0 {
                        return
            }


            for _, volume := range rc.volumesFailedReconstruction {
                        if rc.desiredStateOfWorld.VolumeExistsWithSpecName(volume.podName, volume.volumeSpecName) {
                                    // Some pod needs the volume, don't clean it up and hope that
                                    // reconcile() calls SetUp and reconstructs the volume in ASW.
                                    klog.V(4).InfoS("Volume exists in desired state, skip cleaning up mounts", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
                                    continue
                        }
                        klog.InfoS("Cleaning up mounts for volume that could not be reconstructed", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
                        rc.cleanupMounts(volume)
            }


            klog.V(2).InfoS("Orphan volume cleanup finished")
            // Clean the cache, cleanup is one shot operation.
            rc.volumesFailedReconstruction = make([]podVolume, 0)
}
```

Reply:
I'll update the PR to make it more clear based on the comments.